### ISSUE TITLE
Import maps and performance (HTTP/2)

### ISSUE THREAD

**Issue by nateberkopec** [puma/puma#2697]
Rails is going the way of supporting [import maps by default](https://github.com/rails/importmap-rails) as its new JS solution. This means that Rails apps can make 100+ requests to the application server as they traverse the [import map](https://github.com/WICG/import-maps#the-basic-idea). 

To make this fast in Puma, I have a few concerns:

1. HTTP/2. This is almost certainly a starting requirement. So, we have to decide what approach we're going to take here - graft on helper libraries to the existing native extensions, or scrap the parser entirely and use something in Ruby, etc.
2. Benchmark. Should be easy to create a before/after benchmark here.
3. Rack. I remember last time we evaluated HTTP/2 in #454 there were a lot of concerns about how this would all map to Rack, which were basically never resolved.

Other things we'll probably come up as we're spiking this. Right now, I want to know:

1. Who's interested in getting involved
2. What technical approach we're going to take for HTTP/2 support
3. What other technical barriers might exist (I/O?) for making fast importmaps a reality.

**Comment by nateberkopec** [comment]
> What technical approach we're going to take for HTTP/2 support

Here are the options as I see them:

1. [nghttp2](https://nghttp2.org/). We would need a similar library for Java. This has the disadvantage of pushing more logic into native extensions, and our contributor base in that area is much more limited, so this would end up being a lot of hero-coding by a limited set of people.
2. [http2](https://github.com/igrigorik/http-2) gem by Ilya Grigorik. Advantage: maintained by someone very knowledgable, disadvantage: not sure how this would integrate into existing code around HTTP 1.1.
3. [protocol-http2](https://github.com/socketry/protocol-http2). Advantage of also supporting http/1 via protocol-http1, which means we would get to just rip out all the http-related native extensions.

**Comment by MSP-Greg** [comment]
> This means that Rails apps can make 100+ requests to the application server as they traverse the import map.

A bit confused.  Are 'import maps' kind of like RubyGems for js in the client/browser?  IOW, they're not generating requests to the application server, but to web repositories of the js 'packages'?

Regardless, I'm interested in http/2 and http/3.  Currently (with http/1.1), one  'conversation/request' is communicated to the rack app per socket.  With http/2, one socket can have many streams, and hence, multiple 'conversations/requests' happen on one socket.  How does rack handle that?  I must be missing something, probably need to look at some examples of Ruby http/2 servers...

**Comment by brenogazzola** [comment]
I’m willing to help. Don’t have any experience with app servers, but not afraid of reading code and debugging stuff until I figure what’s happening. 

Worst case, I can at least provide a production app to test a real workload (not necessarily through importmaps, but by having webpack chunk as much as possible)

**Comment by brenogazzola** [comment]
nghttp2 sounds like the least ideal of the three. It feels like it would be the fastest, but a few extra milliseconds per request sound better than bugs being open longer and burned out maintainers. 

protocol2 sounds like a fallback option unless the native extensions are a maintenance problem right now. If they are stable them I’d say there’s no reason to change a winning team. 

http2 is probably the best option. Full ruby and will add the least amount of unknowns to the current code, even if it required a bit extra effort to get it integrated. Silly question from someone who never touched puma code: does it need to integrate with http1 code? Can’t it be an XOR deal? As in “client support http2, so let’s get all requests through the http2 gem code.”?

**Comment by ioquatix** [comment]
Don't underestimate the complexity of HTTP/2.

I'm not sure I'd call the `http2` gem well maintained.

The biggest limitation Puma has in this regard is simply the limited number of threads w.r.t. simultaneous connections.

There is also the reality that most load balancers don't yet support HTTP/2 on the backend.

Falcon supports HTTP/2 with Rack with no changes to applications.

Supporting multiple streams essentially means you'd be implementing a HTTP/2 -> HTTP/1 (application) gateway. You might as well just use Falcon in the connection acceptor thread - and this is only half suggested as a joke. I'm at least somewhat serious. 99% of Falcon is `Async::HTTP`. You could literally pull in Async::HTTP, buffer the entire request, and shoot it off to the Puma worker pool and you'd be able to implement this in about a day and about 100 lines of code.

**Comment by ioquatix** [comment]
By the way, I've often thought of making a "threaded" adaptor for Falcon which executes requests in a thread pool which matches with Rails' expectations on how the execution of web requests works, rather than Falcon/Async which gives you a well defined concurrency model which unfortunately still bumps up against Rails' assumptions about the world ActiveRecord works in. But the reality is, even this is slowly changing. Puma is an incredibly important piece of technology. This probably forms part of a larger conversation but I'm not sure Puma needs HTTP/2. I think there are still improvements to be made to work scheduling and thread pool implementation TBH. I don't see Falcon/Puma as being in competition with each other except on the most friendly terms. They serve different purposes.

**Comment by nateberkopec** [comment]
So it sounds like benchmark would be a good first step then to get an idea of the extent of the issue. That will let us try different configurations, such as putting `h2o` or `nginx` in front of Puma, which will basically be the same perf-wise as Puma support HTTP/2 proper, because as @ioquatix mentioned without a concurrency model change we're basically just implementing a HTTP/2->HTTP/1 gateway.

I want to be clear that the motivation here is not "Puma notionally must support HTTP/2" but "Puma should make the new import-map-driven experience in Rails 7 as fast as possible", so HTTP/2 is just a strong hunch on how we accomplish that. Maybe there are other things we can do. I'm also wondering if, because import maps is primarily about serving files from disk, that maybe there's a shortcut we can take here that makes import maps fast by avoiding Rack entirely.

I think this weekend I'll take some time to make a benchmark and that should provide some insight into next steps, then.

**Comment by byroot** [comment]
> "Puma should make the new import-map-driven experience in Rails 7 as fast as possible"

You mean as a development server, or as a production one? Because I can see the appeal for the former, but for the later, if your assets requests hit your Ruby server, you kind already lost.

And even then, operationally speaking, it doesn't make much sense to multiplex requests all the way to the application server. If you have multiple requests in the same connection, it's much better if your reverse proxy dispatch them to distinct workers. Hence why I don't really see the appeal of HTTP/2 for ruby application servers.

**Comment by nateberkopec** [comment]
definitely mostly focusing on the former, but we can't ignore the prod experience I think either. Puma has always been an app server that you can just throw up on `MY_FAV_VPS` and have a good experience, and if Rails 7 makes that not true, that would be a loss.

**Comment by ioquatix** [comment]
Leveraging HTTP/1 with splice/send file might be sufficient and you could totally build a light weight fiber scheduler for file IO which seems like it would solve any static file serving overhead.

**Comment by nateberkopec** [comment]
So, there appear to be two reasons why this isn't fast today:

1. Server response times for JS in Rails 7 are actually kind of bad. It takes about 60 ms to get to the first byte down the pipe. 
2. The six-connection-per-domain limit (which is a limit imposed by browsers themselves, not in the protocol)

If you take either of those things away, Puma rips just as fast as anything else. HTTP/2 removes the latter limit. I can maybe work with Rails on #1, since 60ms to put a file down the pipe seems kind of bad but I don't know enough about how that works in Rails to know if it will be hard or not.

So that means this problem isn't significantly I/O bound.

[Here is my benchmark setup using k6](https://github.com/nateberkopec/pumaimportmapbench.git). I'm still working on the nginx reverse proxy there to see if an HTTP/2 -> 1 gateway solves the problem.

**Comment by nateberkopec** [comment]
Also, re: TTFB, that will definitely be worse in production because after you set up the initial connection, you still need to make a full roundtrip to ask for the next file you want. So maybe fixing Rails' TTFB here won't do much.

**Comment by nateberkopec** [comment]
<img width="881" alt="Screen Shot 2021-09-19 at 2 59 34 PM" src="https://user-images.githubusercontent.com/845662/133945769-5039a901-0c0d-430b-8cc8-d09cb4c1078b.png">

Example of what the [HOL-blocking](https://en.wikipedia.org/wiki/Head-of-line_blocking) looks like today.

**Comment by brenogazzola** [comment]
If the problem is Sprockets (which is responsible for serving the files), I can help, as I’ve just started spending some time there to fix a couple of bugs. 

We will also have to give some care to the new propshaft gem which will replace sprockets:
https://github.com/rails/propshaft

**Comment by brenogazzola** [comment]
That said. JS and CSS fikes are requested enough that they should have close to 100% cache hit.

**Comment by nateberkopec** [comment]
@brenogazzola Not anymore. [Browser caches are partitioned](https://developers.google.com/web/updates/2020/10/http-cache-partitioning) now. It essentially means the cache key of a request includes the domain name of the current window. 3rd party CDNs will not hit any more often than 1st-party requests.

**Comment by brenogazzola** [comment]
What I get from the article is that the old “use jquery from a cdn because if the user visited another website that has it, it will be already cached” is no longer valid, is that right? 🤔

What I meant is, if your app is using cloudflare, and your users are in US, and you deploy a new js file, puma will only need to serve it 75 times (25 pops * 3 until cache hit) before the requests stop reaching puma. 

Other CDNs have their own rules, but it seems to me the motivation here is “there are going to be many js files now, instead of one, let’s make sure puma can serve them fast”, and it will only matter for those initial 75 requests.

**Comment by ioquatix** [comment]
By only supporting HTTP/1 Puma is in a unique position to serve static files very efficiently using splice/sendfile. I don't know the current implementation, but we should definitely take advantage of it if possible.

**Comment by brenogazzola** [comment]
Nate will know this better than me, but AFAIK Puma is getting the CSS/JS file from sprockets, and both implement the Rack spec, so Sprockets is reading the content of the file and returning it to Puma through the `call` method. No send_file involved?

**Comment by byroot** [comment]
@brenogazzola there's no more sprockets, that's the point.

**Comment by nateberkopec** [comment]
@ioquatix I don't think that's important right now, because as my benchmark shows, HOL-blocking is the bottleneck and not I/O speed.

@brenogazzola Yes, in production, CDNs will alleviate a lot of the load from Puma. However, Puma should be a good and fast experience without any external dependencies, as it has always been. Also, we don't have CDNs in development, and in dev, Puma currently takes about ~8 seconds to fully satisfy the downloads for an import-mapped app with 150 dependencies, which is not a great experience.

**Comment by MSP-Greg** [comment]
Re the 150 dependencies, what type of response bodies are being returned by rack?  Are they enums/arrays, chunked, or respond to `to_path`, etc?

**Comment by brenogazzola** [comment]
> Puma currently takes about ~8 seconds to fully satisfy the downloads for an import-mapped app with 150 dependencies, which is not a great experience.

😱. Ok, I'm convinced, haha

**Comment by byroot** [comment]
> I don't think that's important right now, because as my benchmark shows, HOL-blocking is the bottleneck and not I/O speed.

Well, you serve each individual static file request say 2x faster, it would have a major impact on that HOL-blocking.

I dug a bit on the Rails & Rack code (`ActionDispatch::Static` and `Rack::Files`), and they do `write(read())`. Leveraging `IO.copy_stream` (`sendfile` / `splice` etc) would require quite a bit of work though.

**Comment by nateberkopec** [comment]
@byroot If you make 150 concurrent requests against Puma running Rack::Static (see my benchmark), it completes in less than 100 milliseconds. So I think the latency Rails is adding is coming from somewhere else in the request, not I/O. I can open up `stackprof` and look at it later.

**Comment by byroot** [comment]
I did a quick proof of concept: https://github.com/puma/puma/pull/2703

The current `read(write())`
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     3.37ms    5.89ms  48.28ms   94.46%
    Req/Sec     0.88k   148.97     1.07k    82.08%
  Latency Distribution
     50%    2.21ms
     75%    2.78ms
     90%    4.09ms
     99%   35.75ms
  105651 requests in 1.00m, 108.24GB read
Requests/sec:   1758.39
Transfer/sec:      1.80GB
```

Using `IO.copy_stream`:

```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.06ms    1.09ms  20.98ms   97.94%
    Req/Sec     1.85k   150.69     2.03k    89.92%
  Latency Distribution
     50%    0.94ms
     75%    1.03ms
     90%    1.21ms
     99%    4.91ms
  221380 requests in 1.00m, 226.81GB read
Requests/sec:   3689.18
Transfer/sec:      3.78GB
```

**Comment by ioquatix** [comment]
`IO.copy_stream` will be a fast path in many cases. We will also look to optimise it for the fiber scheduler so you could depend on it if you used async/fiber scheduler in your top level IO reactor - we will offload it to the kernel using splice and io_uring.

Regarding rack, `response.to_path` I believe is an "official" interface for "sendfile" like responses. You can probably just ignore what Rails is doing (don't call `#each`) and literally do the following:

```
length = File.stat(response.to_path).size
peer.write("transfer-encoding: chunked\r\ncontent-length: #{size}\r\n")
IO.copy_stream(response.to_path, peer)
```

Or something like that.? We should really fix this interface. We don't want to buffer everything into a Ruby string if we can help it.

**Comment by byroot** [comment]
> `IO.copy_stream` will be a fast path in many cases.

Yeah, one big downside though is that it doesn't have a proper timeout API. You can wrap it with `Timeout.timeout` and that will work it, but that's far from ideal. There's likely an opportunity to improve it upstream.

> Regarding rack, response.to_path I believe is an "official" interface for "sendfile" like responses.

For delegating to the reverse proxy, yes: https://github.com/rack/rack/blob/d15dd728440710cfc35ed155d66a98dc2c07ae42/lib/rack/sendfile.rb

for the purpose of using `IO.copy_steam`, we could look at either `to_io`, or `respond_to?(:read) || respond_to?(:readpartial)`.

Also I just realized my benchmark isn't quite perfect, `big_response` send a string directly from memory, so it doesn't have to open and read a file, so the difference might be even bigger.

**Comment by MSP-Greg** [comment]
I added big_file.ru to some of the code that's in various PR's, and updated the code in #2696 to use IO.copy_stream.  Results below with the wrk code, with Puma running `-w4 t5:5`, with a 1,074 kB body size:

Master
```
────────wrk────────  ─Request─time─distribution─(ms)─  Worker─requests  ─wrk─requests─
 -t    -c   req/sec   50%    75%    90%    99%   100%  spread   total     total   bad
 20   100     1164     21    383    606    738    816   3.12    17701     17701     0
 26   130     1168     22    531    832   1010   1100   5.26    17730     17730     0
 35   175     1161     22    750   1150   1410   1510   2.29    17675     17675     0
 46   230     1164     23   1010   1550   1900   1990   1.03    17804     17804     0
 60   300     1169     18     20     26   1530   2000   1.96    17926     17926     0
              1165                                    Totals    88836     88836
══════════════════════════════════════════════════════════════════════════════════════
```
Modified PR 2696
```
────────wrk────────  ─Request─time─distribution─(ms)─  Worker─requests  ─wrk─requests─
 -t    -c   req/sec   50%    75%    90%    99%   100%  spread   total     total   bad
 20   100    21178    1.3   21.6   34.6   41.2  263.6   0.21   319947    319947     0
 26   130    21188    1.4   30.4   47.5   55.8  283.8   0.59   320099    320099     0
 35   175    21052    1.6   43.3   67.1   78.2  137.6   0.32   318098    318098     0
 46   230    21201    1.6   59.2   90.6  104.8  168.6   0.45   320772    320772     0
 60   300    21236    1.6   79.8  121.0  138.6  196.6   0.28   321555    321555     0
             21171                                    Totals  1600471   1600471
══════════════════════════════════════════════════════════════════════════════════════
```
Note that the master run also had errors on the last wrk run (-t60 -c300) and the smem data was 'odd'.

Regardless, using `IO.copy_stream` (when the response body supports it) results in a RPS increase from 1,165 to 21,171.  Pretty big increase.  wrk was processing about 21.7 GB per second.

Lastly, I changed the body size to 50 kB, master ran 11,284, PR 2696 ran 23,138.

**Comment by MSP-Greg** [comment]
@byroot

> one big downside though is that it doesn't have a proper timeout API

Is it any different than what the current code uses, which is `IO#each` (same as [`IO#each_line`](https://msp-greg.github.io/ruby_trunk/Core/IO.html#each-instance_method))?

Also, re the data I listed above, `IO.copy_stream` should always be equal to or faster than `IO#each`, but the speed difference is related to how many iterations `#each` does.  That is somewhat indeterminate for most static files, as one would assume they're all compressed?

**Comment by ioquatix** [comment]
`Timeout.timeout` is event-driven if you use the fiber scheduler in Ruby 3.1.

**Comment by byroot** [comment]
> Is it any different than what the current code uses, which is `IO#each` (same as [`IO#each_line`](https://msp-greg.github.io/ruby_trunk/Core/IO.html#each-instance_method))?

I think you are talking about reading the IO the application returned. This one isn't a big concern because we can assume the application is responsible for returning an IO that won't block forever.

The timeout that concerns me is the write one. Right now [the `fast_write` helper uses a timeout](https://github.com/puma/puma/blob/c10d308aa02fa303640574e4a321c757c0fd3b0e/lib/puma/request.rb#L198-L222), if we were to use `copy_stream` it would expose us to DOS attack. e.g. download an 1MB response and read 1B/s, you tied up the puma thread for 1M seconds.

So for a development only feature, or if we're behind a reverse proxy that buffer responses, it would be ok, but other than that it would be a big security risk.

I started looking at adding a timeout to `copy_stream` and so far I think it's possible, but it might take me a long time or just not manage to do it. And at best it would be available in 3.1.

So not sure how puma could make use of it given these constraints.

**Comment by ioquatix** [comment]
We will definitely support timeout with copy_stream on the fiber scheduler if that's any help. However, this might not make it into 3.1.

**Comment by nateberkopec** [comment]
I learned that Gusto's ~7 year old Rails monolith has 1276 javascript assets, with a total size of 31.2 MB (average size of about 24kb). I've modified my benchmark to look similar (it downloads the es shim, at 30kb, 1200 times).

I've done some more testing using [my benchmark](https://github.com/nateberkopec/pumaimportmapbench), and here's what I've learned:

1. Average time per request is down to 7ms. Did something change in Rails here? If so, 👏 I was getting way worse perf before.
1. Puma 5.5.2 with 5 threads can download 1200 JS files of ~30kb in size in 5.7 seconds. Increasing thread count had no measurable effect.
2. Falcon over HTTP/2 with `falcon serve --threaded -n 1` completes in 7.5 seconds (30% slower). Results with `falcon serve --hybrid --forks 1 --threads 25` were the same.
3. Time spent receiving in both servers was about 200µs per request, suggesting little difference (or room to improve) in terms of I/O performance, since just 0.25 seconds is spent actually reading bytes over the entire benchmark.
4. 96% of the benchmark time is spent waiting on Rails for a response.
5. Running additional workers helps quite a bit. With `puma -t 5 -w 4`, the benchmark completes in just 2.5 seconds.

So, locally, Puma seems to do just fine. It looks like any slowness here is caused by Rails' response times, and is not alleviated with HTTP/2. For local performance, we should focus on improving Rails' response times.

For production, it's a different story. I suspect Rails doesn't really care about our story here, because David looks like he's expecting everyone to just use a CDN in prod. That's fine for Rails but like I said above, I would like for Puma to "just work" and provide a decent experience where possible. In production, a single request might take 100-200ms round-trip, which is going to balloon the total benchmark time here into unsustainable territory of like 40 seconds or more.

My benchmark could be improved by actually providing a list of 1200 JS files to download of roughly 30MB total size, rather than just downloading the same file over and over. I think that would show any problems in Rails better, particularly in a cold boot scenario.

**Comment by nateberkopec** [comment]
I'm trying to duplicate Rails' default - 1 process. If Rails wants to increase throughput by increasing process count, they can go ahead and do that.

**Comment by ioquatix** [comment]
I actually have my own benchmarking tools in `async-http` which despite not being as fast as native C code, are pretty decent at assessing differences in concurrency. You can try the gem `benchmark-http` if you are interested.

**Comment by nateberkopec** [comment]
I just tried h2load - a little bit better. 7.3 seconds on NGINX-fronting-Puma, still slower than HTTP1.

**Comment by schneems** [comment]
Hello world. I’m looking at this (but don’t have immediate answers).

At a high level it looks like if the problem is the rails/rack response time and rack will degrade a http2 to http1 that we are limited in any fixes to puma (or any other web server).

Even though supporting http2 would mean that we only have to handle one connection instead of 1200 connections, based on the info here it sounds like the bottleneck isn’t tcp slow start or friends associated but in action dispatch static middleware or somewhere before it.

I am curious where the bulk of time is spent between puma having a parsed request for an asset and writing a response.


I’m curious how well passenger fares here as it serves 

It seems like to get maximum performance we would need both something that understands http2 and doesn’t need to continually open new connections as well as low latency logic to find those assets on disk and serve them.

Which sounds a lot like putting a http2 server in front of your rails app and teaching it how to serve assets which sounds a lot like what passenger does https://www.phusionpassenger.com/library/dev/ruby/rails_integration.html#static-assets-serving. 

Has anyone benched passenger?

**Comment by ioquatix** [comment]
I will revisit this some time later this year and have updated benchmarks for HTTP/2 - right now the overhead of Async::IO is quite decent but using IO::Buffer and direct IO using the fiber scheduler should help a bit (hopefully a lot). I don't see why we can't make this a lot better.

**Comment by ain** [comment]
Valid question by @schneems around Passenger. Perhaps some of these questions have already been answered by the Passenger team?

**Comment by dkniffin** [comment]
Any updates here? I'm migrating a Rails 7 app over to using importmap and I ran into slowness in my test and dev environments, when fetching the many assets. Is there a solution to this problem yet?

**Comment by nateberkopec** [comment]
Thanks for reporting @dkniffin. Asset loading is a fairly complex process, and Puma may or may not be your bottleneck. Since this issue is difficult to reproduce on a simple app and the path forward to improving performance isn't clear here, if you have a half hour and can book some time with me (see CONTRIBUTING.md for link) I'd love to peek at your screen while this is happening so I can get some ideas as to what's going on.

**Comment by BastienL** [comment]
@dkniffin as we were running into the same issue as yours, we set up [Caddy](https://github.com/caddyserver/caddy) in front of Puma to serve the assets in `development` and `test` modes. As Caddy supports HTTP/2, import maps are no longer a problem.

**Comment by dkniffin** [comment]
@nateberkopec Thank you for the offer. I _think_ I found the issue in my slow tests: `config.assets.digest = false`. This config was causing the test environment to not have digests on files, which in turn meant that the files weren't being served up as static files from `public`. That improved a majority of the requests quite a bit.

Another thing was changing relative imports to absolute imports. So `import Foo from "./foo"` to `import Foo from "components/foo"` or whatever. That change also allowed the correct static files to be used.

I still think there would definitely be a benefit from having http/2 support in Rack and puma, but it's less of a concern for me now. (I _think_... still gotta finish getting my test suite passing)

@BastienL Thanks for the tip! I'll definitely check that out.

**Comment by nateberkopec** [comment]
Great, and if anyone else wants to take me up on that offer I'm happy to do it.

You should not need a reverse proxy in front of Puma to enjoy good performance in development, so I'm not happy with that as a solution, only as a workaround.

**Comment by navidemad** [comment]
To make it work with https://localhost, as mentioned by @BastienL caddy reverse proxy v2
```
curl -sS https://webi.sh/caddy | sh
caddy reverse-proxy --from localhost --to :3000
``` 
Unfortunately, I couldn't found out a way to make it works in my Github Action CI.
It would be awesome to have it out of the box in puma 🥇 I look forward to seeing that 😃

**Comment by darinwilson** [comment]
For others like me who stumbled onto this issue while looking into using import maps and puma: it sounds like David might be planning on releasing something to help bridge the current gap: https://twitter.com/FORSBERGtwo/status/1736766444485099794

**Comment by justinko** [comment]
https://github.com/oesmith/puffing-billy can also aid in test env

**Comment by justinko** [comment]
maybe puma-dev could utilize thruster

**Comment by KonnorRogers** [comment]
Chiming in here that as of Jan 1st with importmap-rails 2.0, importmaps are now always vendored with no option to use a CDN. So on the off chance someone doesn't have an HTTP/2 reverse proxy in front of their Rails app, they may hit some performance dropoff.

I don't know much about Rack or Puma, but I know a lot about importmaps and am happy to help.

@nateberkopec If I can get a reproduction together of a new Rails app with importmaps with a  that times out either in test or dev, or is slow enough to be noticeably causing delays, would that help?

**Comment by taylorthurlow** [comment]
The thing that @darinwilson said DHH may be cooking up is probably [basecamp/thruster](https://github.com/basecamp/thruster), just released as part of Rails World and Kamal 2.0. It doesn't really address the lack of HTTP/2 in development, but it seems clear at this point that the expectation from the Rails side of things is that you run your production app behind a reverse proxy with HTTP/2 support like thruster.

**Comment by northeastprince** [comment]
From what I saw Thruster was more of a hack than a permanent solution. You shouldn't have to have even a small proxy in front of Puma to deploy to production.

**Comment by tcannonfodder** [comment]
For folks that have been using `puma-dev` and have been running into this HTTP/2 performance issue, I've started working on a "rewrite" that uses `caddy`: https://github.com/practical-computer/caddy-dev

It's just a single-line shell script and a `dnsmasq` + resolver configuration. It's very early stages, but I'm going to run it as my daily driver for a month or so then get a proper script for easier installation ready
